### PR TITLE
Evict TiKV leader before resize PVCs

### PR DESCRIPTION
**Pull request by KanShiori**
### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 

1. Deploy cluster with three TiKV.
2. Scale up the storage of TiKV spec.
3. Check the `ComponentVolumeResizing` condition existing in TiKV status.
4. Update the image of TiKV spec, check upgrade is skipped.
5. For three PVCs and Pods, check:
    1. The `tidb.pingcap.com/evict-leader-for-resize: none` annotation in resized Pods is removed 
    1. The `tidb.pingcap.com/evict-leader-for-resize: none` annotation is added to the current Pod.
    2. The storage request is updated of the PVC after the leader count in `status.stores` become 0.
    2. Update the status of the PVC to mock PVC is resized.
    3. Check the `status.volumes` updated in TiKV status and  the `BeginResize` condition existing.
6. After all volume are resized, check all Pods don't have the `tidb.pingcap.com/evict-leader-for-resize: none` annotation.
7. Check the `ComponentVolumeResizing` condition removed in TiKV status.

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Evict TiKV leader before resize PVCs
```

### CODE REVIEW

File: tools/tools.go [hunk 1]
```diff
@@ -11,6 +11,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//go:build tools
```
Comment (DanielZhangQD) [inline tools/tools.go:14]: Is this required?

File: pkg/manager/member/pvc_resizer.go [hunk 2]
```diff
@@ -78,68 +81,69 @@ var (
 	dmWorkerRequirement = util.MustNewRequirement(label.ComponentLabelKey, selection.Equals, []string{label.DMWorkerLabelVal})
 )
 
-type pvcResizer struct {
-	deps *controller.Dependencies
+type volumePhase string
+
+const (
+	// needResize means the storage request of PVC is less than the storage request in TC/DC.
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:87]:
```suggestion
	// needResize means the storage request of PVC is different from the storage request in TC/DC.
```

File: pkg/manager/member/pvc_resizer.go [hunk 3]
```diff
@@ -78,68 +81,69 @@ var (
...
+	needResize volumePhase = "NeedResize"
+	// needResize means the storage request of PVC is equal to the storage request in TC/DC And PVC is resizing.
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:89]:
```suggestion
	// resizing means the storage request of PVC is equal to the storage request in TC/DC and PVC is resizing.
```

File: pkg/manager/member/pvc_resizer.go [hunk 4]
```diff
@@ -78,68 +81,69 @@ var (
...
+	resizing volumePhase = "Resizing"
+	// needResize means the storage request of PVC is equal to the storage request in TC/DC And PVC is resized.
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:91]:
```suggestion
	// resized means the storage request of PVC is equal to the storage request in TC/DC and PVC has been resized.
```

File: pkg/manager/member/pvc_resizer.go [hunk 5]
```diff
@@ -385,132 +375,284 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
+	logPrefix := fmt.Sprintf("before resizing volumes of Pod %s/%s for %q", resizePod.Namespace, resizePod.Name, ctx.ComponentID())
+
+	switch ctx.status.GetMemberType() {
+	case v1alpha1.TiKVMemberType:
+		tc, ok := ctx.cluster.(*v1alpha1.TidbCluster)
+		if !ok {
+			return fmt.Errorf("%s: cluster is not tidb cluster", logPrefix)
+		}
+
+		if !tc.TiKVAllStoresReady() {
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:581]: No need to check all stores ready, is it OK to check that the current store is ready after resizing, and then we can go to scale the next one?

File: pkg/manager/member/pvc_resizer.go [hunk 6]
```diff
@@ -385,132 +375,284 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
+			return fmt.Errorf("%s: any store is not ready", logPrefix)
+		}
+
+		// remove leader eviction ann from reized pod
+		for _, podVolume := range ctx.actualPodVolumes {
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:586]: Why remove the anno before resizing? We should do that after the resizing is done, right?

File: pkg/manager/member/pvc_resizer.go [hunk 7]
```diff
@@ -575,10 +720,30 @@ func (p *pvcResizer) isVolumeExpansionSupported(storageClassName string) (bool,
...
+	_, exist := pod.Annotations[v1alpha1.EvictLeaderAnnKeyForResize]
+
+	if need && !exist {
+		if pod.Annotations == nil {
+			pod.Annotations = map[string]string{}
+		}
+		pod.Annotations[v1alpha1.EvictLeaderAnnKeyForResize] = v1alpha1.EvictLeaderValueNone
+		_, err := client.CoreV1().Pods(pod.Namespace).Update(context.TODO(), pod, metav1.UpdateOptions{})
+		if err != nil {
+			return false, fmt.Errorf("add ann %s to pod failed: %s", v1alpha1.EvictLeaderAnnKeyForResize, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:733]: Add Pod info

File: pkg/manager/member/pvc_resizer.go [hunk 8]
```diff
@@ -575,10 +720,30 @@ func (p *pvcResizer) isVolumeExpansionSupported(storageClassName string) (bool,
...
+		}
+		return true, nil
 	}
+	if !need && exist {
+		delete(pod.Annotations, v1alpha1.EvictLeaderAnnKeyForResize)
+		_, err := client.CoreV1().Pods(pod.Namespace).Update(context.TODO(), pod, metav1.UpdateOptions{})
+		if err != nil {
+			return false, fmt.Errorf("remove ann %s of pod failed: %s", v1alpha1.EvictLeaderAnnKeyForResize, err)
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:741]: Add Pod info

File: pkg/manager/member/pvc_resizer.go [hunk 9]
```diff
@@ -385,132 +375,284 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
+				return fmt.Errorf("remove leader eviction annotation from pod failed: %s", err)
+			}
+			if updated {
+				klog.Infof("end resizing for %s: remove leader eviction annotation from pod %s/%s",
+					ctx.ComponentID(), podVolume.pod.Namespace, podVolume.pod.Name)
+			}
 		}
 	}
 
+	ctx.status.RemoveCondition(v1alpha1.ComponentVolumeResizing)
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:654]: Is it better to just set the condition to false instead of removing it?

Comment (DanielZhangQD) [reply pkg/manager/member/pvc_resizer.go:654]: Then we also have the timestamp (last transition timestamp) when the last resizing is done.

File: pkg/manager/member/pvc_resizer.go [hunk 10]
```diff
@@ -385,132 +375,290 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
+		if !condExist {
+			return p.beginResize(ctx)
+		}
 
-		for volName, pvc := range podVolume.volToPVCs {
-			pvcID := fmt.Sprintf("%s/%s", pvc.Namespace, pvc.Name)
+		// some volumes need to be resized
+		if len(classifiedVolumes[needResize]) != 0 {
+			klog.V(4).Infof("start to resize volumes of Pod %s/%s for %s", resizingPod.Namespace, resizingPod.Name, ctx.ComponentID())
+			return p.resizeVolumesForPod(ctx, resizingPod, classifiedVolumes[needResize])
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:429]: This is pod-specific logic, can we separate it from the condition update logic which is component-level logic?

File: pkg/manager/member/pvc_resizer.go [hunk 11]
```diff
@@ -385,132 +375,290 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
-				klog.Errorf("Check PVC %q of %q resized failed: storage request is empty", pvcID, id)
-				continue
-			}
-			currentCapacity, ok := pvc.Status.Capacity[corev1.ResourceStorage]
-			if !ok {
-				klog.Errorf("Check PVC %q of %q resized failed: storage capacity is empty", pvcID, id)
-				continue
-			}
+		// some volumes are resizing
+		for _, volume := range classifiedVolumes[resizing] {
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:433]: ditto.

File: pkg/manager/member/pvc_resizer.go [hunk 12]
```diff
@@ -385,132 +375,290 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
+	allResized := resizingPod == nil
+	condExist := meta.IsStatusConditionTrue(ctx.status.GetConditions(), v1alpha1.ComponentVolumeResizing)
 
-		allResized := true
-		errors := []error{}
+	if allResized {
+		if condExist {
+			return p.endResize(ctx)
+		}
+		klog.V(4).Infof("all volumes are resized for %s", ctx.ComponentID())
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:420]: Is ti OK to return after this log?

File: pkg/manager/member/pvc_resizer.go (hunk 6 above)
Comment (DanielZhangQD) [reply pkg/manager/member/pvc_resizer.go:583]: Is it more readable that we check the store status UP and remove the annotation before we decide to scale PVCs for another Pod in L400 block?

File: pkg/manager/member/pvc_resizer.go [hunk 13]
```diff
@@ -385,132 +375,290 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
-				errors = append(errors, err)
-				continue
+				return err
+			}
+			if updated {
+				klog.Infof("%s: remove leader eviction annotation from pod %s/%s", logPrefix, pod.Namespace, pod.Name)
+			}
+			for _, store := range tc.Status.TiKV.Stores {
+				if store.PodName == pod.Name && store.State != v1alpha1.TiKVStateUp {
+					return fmt.Errorf("%s: store %s is not ready", logPrefix, store.ID)
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:598]: Add Pod info

File: pkg/manager/member/pvc_resizer.go [hunk 14]
```diff
@@ -385,132 +375,290 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
 		}
 	}
 
+	ctx.status.SetCondition(metav1.Condition{
+		Type:    v1alpha1.ComponentVolumeResizing,
+		Status:  metav1.ConditionFalse,
+		Reason:  "EndResizing",
+		Message: "All volumes are resized",
+	})
+	klog.Infof("end resizing for %s: remove resizing condition", ctx.ComponentID())
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:662]: Needs update

File: pkg/manager/member/pvc_resizer.go [hunk 15]
```diff
@@ -385,132 +375,290 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
+		if len(curClassifiedVolumes[resizing]) != 0 || len(curClassifiedVolumes[needResize]) != 0 {
+			resizingPod = podVolumes.pod
+			classifiedVolumes = curClassifiedVolumes
+			break
+		}
+	}
 
-	for _, podVolume := range podVolumes {
+	allResized := resizingPod == nil
+	condExist := meta.IsStatusConditionTrue(ctx.status.GetConditions(), v1alpha1.ComponentVolumeResizing)
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:414]:
```suggestion
	condResizing := meta.IsStatusConditionTrue(ctx.status.GetConditions(), v1alpha1.ComponentVolumeResizing)
```

File: pkg/manager/member/pvc_resizer.go (hunk 13 above)
Comment (KanShiori) [reply pkg/manager/member/pvc_resizer.go]: Done

File: pkg/manager/member/pvc_resizer.go (hunk 14 above)
Comment (KanShiori) [reply pkg/manager/member/pvc_resizer.go]: Done

File: pkg/manager/member/pvc_resizer.go (hunk 15 above)
Comment (KanShiori) [reply pkg/manager/member/pvc_resizer.go]: Done

File: pkg/manager/member/pvc_resizer.go [hunk 16]
```diff
@@ -385,132 +375,291 @@ func (p *pvcResizer) updateVolumeStatus(ctx *componentVolumeContext) {
...
-						pvcID, id, currentRequest.String(), currentCapacity.String())
-				} else {
-					klog.V(4).Infof("PVC %q of %q is already resized, request: %s",
-						pvcID, id, quantityInSpec.String())
-				}
-				continue
+	// some volumes are resizing
+	for _, volume := range classifiedVolumes[resizing] {
+		klog.Infof("PVC %s/%s for %s is resizing", volume.pvc.Namespace, volume.pvc.Name, ctx.ComponentID())
+	}
```
Comment (DanielZhangQD) [inline pkg/manager/member/pvc_resizer.go:437]: Should we move this to L427?

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 17]
```diff
@@ -1201,7 +1217,16 @@ type TiDBFailureMember struct {
...
 
-const EvictLeaderAnnKey = "tidb.pingcap.com/evict-leader"
+var (
+	EvictLeaderAnnKeys = []string{EvictLeaderAnnKey, EvictLeaderAnnKeyForResize}
+)
+
+const (
+	// EvictLeaderAnnKey is the annotation key to evict leader used by user.
+	EvictLeaderAnnKey = "tidb.pingcap.com/evict-leader"
+	// EvictLeaderAnnKey is the annotation key to evict leader user by pvc resizer.
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1227]:
```suggestion
	// EvictLeaderAnnKeyForResize is the annotation key to evict leader user by pvc resizer.
```

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 18]
```diff
@@ -376,6 +376,14 @@ const (
 	TidbClusterReady TidbClusterConditionType = "Ready"
 )
 
+// ComponentConditionType represents a component condition value.
+type ComponentConditionType string
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/types.go:380]: is this type used?

File: pkg/controller/tidbcluster/pod_control.go [hunk 19]
```diff
@@ -340,3 +340,14 @@ func (c *PodController) syncTiKVPod(ctx context.Context, pod *corev1.Pod, tc *v1
 
 	return reconcile.Result{}, nil
 }
+
+func needEvictLeader(pod *corev1.Pod) (string, string, bool) {
+	for _, key := range v1alpha1.EvictLeaderAnnKeys {
+		value, exist := pod.Annotations[key]
+		if exist {
+			return key, value, true
```
Comment (csuzhangxc) [inline pkg/controller/tidbcluster/pod_control.go:348]: any restriction to multi evict keys (with differet values)?

Comment (KanShiori) [reply pkg/controller/tidbcluster/pod_control.go:348]: no, just use a default priority.

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 18 above)
Comment (KanShiori) [reply pkg/apis/pingcap/v1alpha1/types.go]: no, removed.

**Review by csuzhangxc — APPROVED** [review]

**Pull request merged**